File: python/64_Minimum_Path_Sum.py
class Solution:
    # @param {integer[][]} grid
    # @return {integer}
    def minPathSum(self, grid):
        if not grid:
            return 0
        row=len(grid)
        col=len(grid[0])
        below=[0]*(col+1)
        cur=[0]*col
        for i in range(col-1,-1,-1):
            below[i]=below[i+1]+grid[row-1][i]
        if row==1:
            return below[0]
        for r in range(row-2,-1,-1):
            for c in range(col-1,-1,-1):
                if c==col-1:
                    cur[c]=below[c]+grid[r][c]
                else:
                    cur[c]=min(below[c],cur[c+1])+grid[r][c]
            below=cur
        return cur[0]

File: Java/64_Minimum_Path_Sum.java
public class Solution {
    public int minPathSum(int[][] grid) {
        // similar to #62 #63
        if(grid==null || grid.length==0||grid[0].length==0)return 0;
        int m = grid.length;
        int n = grid[0].length;
        int[] pre = new int[n];
        int[] cur = new int[n];
        int i,j;
        for(i=0;i<n;i++){
            pre[i]= i==0?grid[0][i]:pre[i-1]+grid[0][i];
        }
        
        for(i=1;i<m;i++){
            for(j=0;j<n;j++){
                cur[j]=grid[i][j];
                if(j>0){
                    cur[j]+=Math.min(pre[j],cur[j-1]);
                }else{
                    cur[j]+=pre[j];
                }
            }
            for(j=0;j<n;j++){
                pre[j]=cur[j];
                cur[j]=0;
            }
        }
        return pre[n-1];
    }
}